app/controllers/add_user.js

```diff
@@ -0,0 +1,281 @@
+    if (this.get("organisationId")) {
+      let store = this.get("store");
+      let organisation = store.peekRecord(
+        "gc_organisation",
+        this.get("organisationId")
+      );
+      this.set("organisationName", organisation && organisation.get("nameEn"));
+    }
+  }),
+
+  clearFormData() {
+    this.set("firstName", "");
+    this.set("lastName", "");
+    this.set("mobilePhone", "");
+    this.set("email", "");
+    this.set("selectedRoleIds", []);
+    this.set("disabled", false);
+    this.set("organisationName", "");
+    this.set("newUploadedImage", null);
+  },
+
+  formatMobileNumber() {
+    const mobile = this.get("mobilePhone");
+    if (mobile.length) {
+      return config.APP.HK_COUNTRY_CODE + mobile;
+    }
+  },
+
+  getRequestParams() {
+    const mobilePhone = this.formatMobileNumber();
+
+    return {
+      firstName: this.get("firstName"),
+      lastName: this.get("lastName"),
+      mobile: mobilePhone,
+      email: this.get("email"),
+      disabled: this.get("disabled"),
+      organisations_users_ids: [this.get("organisationId")]
+    };
+  },
+
+  saveImage() {
+    let image = this.get("newUploadedImage");
+    if (image) {
+      return image.save();
+    } else {
+      return Promise.resolve();
+    }
+  },
+
+  actions: {
+    searchOrganization() {
+      this.replaceRoute("search_organisation", {
+        queryParams: {
+          backToNewUser: true
+        }
+      });
+    },
+
+    setSelecteIds(id, isSelected) {
+      if (isSelected) {
+        this.get("selectedRoleIds").pushObject(id);
+      } else {
+        this.get("selectedRoleIds").removeObject(id);
+      }
+
+      let hasCharityRole =
+        this.get("selectedRoleIds").indexOf(this.get("charityRoleId")) >= 0;
+      this.set("displayOrganizationInput", hasCharityRole);
+    },
+
+    saveUser() {
+      let _this = this;
+      const loadingView = getOwner(_this)
+        .lookup("component:loading")
+        .append();
+
+      let newUser = _this
+        .get("store")
+        .createRecord("user", _this.getRequestParams());
+
+      if (_this.get("selectedRoleIds.length")) {
+        newUser.set("userRoleIds", _this.get("selectedRoleIds"));
+      } else {
+        newUser.set("userRoleIds", []);
+      }
+
+      _this.saveImage().then(imageData => {
+        if (imageData) {
+          let image = _this.get("store").peekRecord("image", imageData.id);
+          newUser.set("image", image);
+        }
+
+        newUser
+          .save()
+          .then(data => {
+            this.clearFormData();
+            this.transitionToRoute("user_details", data.id);
+          })
+          .catch(xhr => {
+            if (xhr.status === 422) {
+              this.get("messageBox").alert(xhr.responseJSON.errors[0].message);
+            } else {
+              throw xhr;
+            }
+          })
+          .finally(() => loadingView.destroy());
+      });
+    },
```

Comment:
Using the `async` mixin we have, and an async method you could significantly reduce the size of a lot of those methods, making them more readable. 

As an example:

```javascript
  saveUser() {
    let newUser = this
      .get("store")
      .createRecord("user", this.getRequestParams());

    newUser.set("userRoleIds", this.getWithDefault("selectedRoleIds", []));

    return this.runTask(async () => {
      newUser.set("image", await this.saveImage());

      const { id } = await newUser.save();

      this.clearFormData();
      this.transitionToRoute("user_details", id);

    }, this.ERROR_STRATEGIES.MODAL);
  }
```

we basically don't want manually creating the spinner anywhere, or the error modal.

Note: doing `let _this = this` is unnecessary, lambda methods use the current scope